libs/commons/List_.ml

```diff
@@ -222,6 +222,16 @@ let tl_exn errmsg xs =
   | [] -> failwith errmsg
   | _ :: tail -> tail
 
+let rec last_exn errmsg xs =
+  match xs with
+  | [] -> failwith errmsg
+  | [ x ] -> x
+  | _ :: tl -> last_exn errmsg tl
+
+let last_opt xs =
+  try Some (last_exn "last_opt failed" xs) with
+  | Failure _ -> None
```

Comment:
I'd rather not encourage people to use exceptions. Just implement last_opt directly and leave last_exn out.